In [2]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression as skLinearRegression
from sklearn.metrics import mean_squared_error
from sklearn import datasets



In [3]:
# Model implementation
class LinearRegression:

    def __init__(self,alpha = 0.001, n_iters = 1000):
        self.alpha = alpha
        self.n_iters = n_iters
        self.weights = None
        self.bias = None

    def fit(self,X,y):
        m , n = X.shape 
        self.weights = np.zeros(n)
        self.bias = 0

        for _ in range(self.n_iters):
            # model
            y_pred = np.dot(self.weights,X.T) +self.bias
            # gradients 
            dj_dw = (1/m) * np.dot((y_pred- y),X)
            dj_db = (1/m) * np.sum(y_pred- y)

            self.weights -= self.alpha * dj_dw
            self.bias -= self.bias * dj_db







    def predict(self, X):
        y_pred = np.dot(self.weights,X.T) +self.bias
        return y_pred
    


    
    def mse(self,y_test , y_pred):
        return np.mean((y_test - y_pred)**2)


In [4]:
# Model testing
X, y = datasets.make_regression(n_samples=100,n_features= 1,noise=20 , random_state=4)
X_train , X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=1234)


reg = LinearRegression(alpha=0.01 , n_iters=500)
reg.fit(X_train,y_train)
pred = reg.predict(X_test)
mse = reg.mse(y_test,pred)

print(f"mse = {mse}")


mse = 305.7455031898213


In [5]:
#sklearn model
model = skLinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)

print(f"mse = {mse}")

mse = 305.7741316228642


## Proof of concept
my model achieved an MSE of 305.75 , while the scikit-learn model achieved an MSE of 305.77 . The small difference in MSE between my model and the scikit-learn model suggests that my model is reasonably accurate and performs similarly to the industry standard. However, it's important to note that the accuracy of my model may vary depending on the specific data it is trained on and the features used. Further experimentation and testing may be necessary to determine the true accuracy of my model in different contexts.